// @sap/cds-dk/bin/upgrade.js
module.exports = Object.assign(upgrade, {
    init: cmd => new (require('../lib/client/tenant_action'))(cmd),
    options: ['--at', '--user'],
    flags: ['--local'],
    shortcuts: ['-a', '-u', '-l'],
    help: `
# SYNOPSIS

    *cds upgrade* <tenant>

    Upgrade a tenant subscribed to a multitenant SaaS app to its latest base
    model.

    Running *cds upgrade* with a URL is intended for non-productive use only,
    including against SaaS apps running on localhost with mock authentication.

# OPTIONS

    *-a* | *--at* <url>

        Specify the URL of the SaaS app. Determined from the current user's
        active CDS or Cloud Foundry login by default.

    *-u* | *--user* <name>[:[<password>]]

        Username and optionally password for authentication with Basic Auth in
        test scenarios.
        
        Only relevant if <url> is given. May be omitted in case of dummy
        authentication.
        
        Tokens saved with *cds login* are not used for this command.

    *-l* | *--local*

        Run a locally-installed @sap/cds-mtxs and connect to it instead of a
        URL. Ignore settings saved with *cds login*.


# SEE ALSO

    *cds login* for more information on authentication options.
    
    *cds login* to save project settings and Basic-Auth data, simplifying multiple runs of this command.

`});


async function upgrade([tenant], options = {}) {
    const { getMessage } = require('../lib/client/helper/logging');
    if (!tenant) {
        throw getMessage('Tenant not given.', { command: 'upgrade' });
    }
    const libOptions = { tenant };
    if (options.local) {
        libOptions.local = true;
        delete options.local;
        if (options.user) {
            throw getMessage('Username cannot be specified when running @sap/cds-mtxs locally.', { command: 'upgrade' });
        }
        if (options.at) {
            throw getMessage('URL cannot be specified when running @sap/cds-mtxs locally.', { command: 'upgrade' });
        }
    }
    const [username, password] = options.user?.split(':') ?? [];
    delete options.user;

    await upgrade.init('upgrade').run({
        ...options,
        username,
        password
    }, libOptions);
}
