// @sap/cds-dk/lib/init/template/redis-messaging/index.js
const { join } = require('path')
const cds = require('../../../cds')
const { read } = cds.utils
const { readProject } = require('../../util/projectReader')
const { mergeJSON, mergeYAML } = require('../../util/merge')
const { srvNode4, srvJava4, redis } = require('../_merging/registry-mta')

module.exports = class RedisMessagingTemplate extends require('../templateBase') {

  async canRun() {
    const { hasMta, hasHelm, isJava } = await readProject()
    if (isJava)             throw `'cds add redis-messaging' is not available for Java yet`
    if (hasHelm && !hasMta) throw `'cds add redis-messaging' is not available for Kyma yet`
    return true
  }

  static hasFacet(env) {
    return env.requires?.messaging?.kind === 'redis-messaging'
  }

  async run() {
    const project = await readProject()
    const { configFile } = project
    project.shortcut = !(await read(configFile)).cds?.requires?.messaging?.kind
    await mergeJSON(configFile, join(__dirname, 'files', 'package.json.hbs'), project)
  }

  async runDependentMerging() {
    const project = await readProject()
    const { hasMta, isNodejs, srvPath } = project

    if (hasMta) {
      const srv = (isNodejs ? srvNode4 : srvJava4)(srvPath)
      await mergeYAML(
          'mta.yaml',
          join(__dirname, 'files', 'mta.yaml.hbs'),
          project,
          {
              additions: [srv, redis],
              relationships: [{
                  insert: [redis, 'name'],
                  into: [srv, 'requires', 'name']
              }]
          }
      )
    }
  }
}
