// @sap/cds-dk/lib/init/template/notifications/index.js
const { join } = require('path')
const { readProject } = require('../../util/projectReader')
const { mergeJSON, mergeYAML } = require('../../util/merge')
const { DESTINATION } = require('../../constants').OPTIONS

const { notificationTypesDeployer } = require('../_merging/registry-mta')

module.exports = class NotificationsTemplate extends require('../templateBase') {

  static hasFacet(env) {
    return !!env.requires?.notifications;
  }

  async canRun() {
    const project = await readProject()
    const { isJava } = project
    if (isJava) {
      throw `'cds add notifications' is not available for Java yet`
    }
    return true
  }

  getDependencies() {
    return [DESTINATION]
  }

  async run() {
    const project = await readProject()
    const { configFile } = project
    await mergeJSON(configFile, join(__dirname, 'files', 'package.json.hbs'), project)
  }

  async runDependentMerging() {
    const project = await readProject()
    const { hasMta } = project

    if (hasMta) {
      await mergeYAML('mta.yaml', join(__dirname, 'files', 'mta.yaml.hbs'), project, {
        additions: [notificationTypesDeployer]
      })
    }
  }
}
